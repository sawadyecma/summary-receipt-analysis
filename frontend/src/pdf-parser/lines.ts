import { isTextItem, TextItemsFromTextContent } from "../pdfjs-helper/helper";

export type Lines = Line[];

export type Line = {
  body: string[];
};

export const composeLinesFromTextItems = (
  items: TextItemsFromTextContent
): Lines => {
  let tempLine: string[] = [];
  let ret: Lines = [];
  items.forEach((item) => {
    if (isTextItem(item)) {
      tempLine.push(item.str);

      if (item.hasEOL) {
        tempLine = [];
        ret.push({
          body: tempLine,
        });
      }
    }
  });
  return ret;
};
