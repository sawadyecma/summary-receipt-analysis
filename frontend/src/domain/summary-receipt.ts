import { Line, Lines } from "../pdf-parser/lines";

export type SummaryReceipt = {
  body: (SectionHeader | Session | Break | Total)[];
};

export const SESSION_TYPE = {
  header: "header",
  session: "session",
  break: "break",
  totals: "totals",
} as const;

type SessionType = (typeof SESSION_TYPE)[keyof typeof SESSION_TYPE];

type SectionHeader = {
  sectionType: typeof SESSION_TYPE.header;
  body: string[];
};

type Session = {
  sectionType: typeof SESSION_TYPE.session;
  body: string[];
};

type Break = {
  sectionType: typeof SESSION_TYPE.break;
  // breakType: "paid-break" | "unpaid-break";
  body: string[];
};

type Total = {
  sectionType: typeof SESSION_TYPE.totals;
  body: string[];
};

const lineJudger = () => {
  const treatAsTitle = (line: Line): string => {
    const joined = line.body.join("").trim();
    return joined;
  };

  const isSessionTitle = (line: Line): boolean => {
    return treatAsTitle(line) === "Session";
  };

  const isBreakTitle = (line: Line): boolean => {
    return treatAsTitle(line) === "Break";
  };
  const isTotalTitle = (line: Line): boolean => {
    return treatAsTitle(line) === "Totals";
  };

  const isBeginningOfSection = (line: Line) => {
    return isBreakTitle(line) || isSessionTitle(line) || isTotalTitle(line);
  };

  return { isSessionTitle, isBreakTitle, isBeginningOfSection, isTotalTitle };
};

export const composeSummaryReceipt = (lines: Lines): SummaryReceipt => {
  const { isSessionTitle, isBreakTitle, isBeginningOfSection, isTotalTitle } =
    lineJudger();

  const ret: SummaryReceipt = {
    body: [],
  };
  const header: SectionHeader = {
    sectionType: "header",
    body: [],
  };

  let nextSectionType: SessionType = SESSION_TYPE.header;

  let currentSection: SummaryReceipt["body"][number] = header;

  lines.forEach((line, index, arr) => {
    currentSection.body.push(line.body.join(""));

    const nextIndex = index + 1;
    const nextLine = nextIndex >= lines.length ? undefined : arr[nextIndex];
    if (!nextLine) {
      ret.body.push(currentSection);

      return;
    }

    if (isBeginningOfSection(nextLine)) {
      // update currentSection
      // push currentSection
      // make next section with empty
      if (isSessionTitle(nextLine)) {
        nextSectionType = SESSION_TYPE.session;
      } else if (isBreakTitle(nextLine)) {
        nextSectionType = SESSION_TYPE.break;
      } else if (isTotalTitle(nextLine)) {
        nextSectionType = SESSION_TYPE.totals;
      }
      ret.body.push(currentSection);
      const emptyBody: string[] = [];

      currentSection = {
        sectionType: nextSectionType,
        body: emptyBody,
      };
    }
  });

  return ret;
};
