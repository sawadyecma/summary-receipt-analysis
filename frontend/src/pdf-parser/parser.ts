import * as PDFJS from "pdfjs-dist";
import pdfWorker from "pdfjs-dist/build/pdf.worker.mjs?url";

import { isTextItem } from "../pdfjs-helper/helper";
import { composeLinesFromTextItems, Lines } from "./lines";
import { extractOneFileFromEvent } from "../util/dom-util";

PDFJS.GlobalWorkerOptions.workerSrc = pdfWorker;

export const composeFileChangeHandler = (
  callback: (result: string, lines: Lines) => void
) => {
  const fileChangeHandler = async (ev: HTMLElementEventMap["change"]) => {
    const extracted = extractOneFileFromEvent(ev);
    if (extracted.error) {
      alert(extracted.error.message);
      return;
    }

    const file = extracted.result;

    const arrayBuffer = await file.arrayBuffer();
    const doc = PDFJS.getDocument(arrayBuffer);
    const pdf = await doc.promise;

    const numPages = pdf.numPages;
    console.log({ numPages });

    // NOTE: The first page is 1
    for (let i = 1; i <= numPages; i++) {
      console.log({ i });
      const page = await pdf.getPage(i);
      const tree = await page.getStructTree();
      const text = await page.getTextContent();

      console.log({ page, tree });
      console.log({ text });

      let res = "";
      text.items.forEach((item) => {
        if (isTextItem(item)) {
          res += item.str;
        }
      });
      const lines = composeLinesFromTextItems(text.items);
      console.log({ lines });

      callback(res, lines);
    }
  };

  return { fileChangeHandler };
};
