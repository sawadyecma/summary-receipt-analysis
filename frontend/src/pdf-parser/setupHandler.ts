import { composeSummaryReceipt, SESSION_TYPE } from "../domain/summary-receipt";
import { composeFileChangeHandler } from "./parser";

export const setupSummaryReceiptParser = (
  inputEle: HTMLInputElement,
  resultEle: HTMLElement
) => {
  const { fileChangeHandler } = composeFileChangeHandler((result, lines) => {
    const normalized = composeSummaryReceipt(lines);

    const sectionColor = {
      [SESSION_TYPE.header]: "#D1603D",
      [SESSION_TYPE.session]: "#D0E37F",
      [SESSION_TYPE.break]: "#DDB967",
      [SESSION_TYPE.totals]: "#D1603D",
    } as const;

    const normalizedDom = normalized.body
      .map((section) => {
        const joined = section.body.join("<br/>");

        const color = sectionColor[section.sectionType];

        const style = `
          border: solid 2px ${color};
          padding: 8px 16px 4px;
        `;

        return `<p style='${style}' >${joined}</p>`;
      })
      .join("");

    resultEle.innerHTML = normalizedDom;
    result;
  });

  inputEle.addEventListener("change", fileChangeHandler);
};
