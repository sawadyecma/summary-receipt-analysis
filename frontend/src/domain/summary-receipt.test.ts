import { describe, expect, test } from "vitest";
import { Lines } from "../pdf-parser/lines";
import { composeSummaryReceipt } from "./summary-receipt";

const testData: { lines: Lines } = {
  lines: [
    {
      body: ["Name:", " ", "Bubbles,", " ", "Jackson"],
    },
    {
      body: ["Employee", " ", "ID:", " ", "126167"],
    },
    {
      body: ["Pick", " ", "Date:", " ", "11/02/2025"],
    },
    {
      body: ["Start", " ", "Time:", " ", "11/02/2025", " ", "06:30"],
    },
    {
      body: ["End", " ", "Time:", " ", "11/02/2025", " ", "12:02"],
    },
    {
      body: ["Attendance", " ", "Hrs:", " ", "5.53", ""],
    },
    {
      body: ["Session", ""],
    },
    {
      body: ["Location:", " ", "[Bengeo]608"],
    },
    {
      body: ["Variety:", " ", "Victoria"],
    },
    {
      body: ["Pay", " ", "Type:", " ", "Per", " ", "Unit"],
    },
    {
      body: ["Unit:", " ", "125g", " ", "Driscoll's"],
    },
    {
      body: ["Piece", " ", "Rate:", " ", "0.71"],
    },
    {
      body: ["Target", " ", "Pick", " ", "Rate:", " ", "48.00"],
    },
    {
      body: [
        "Session",
        " ",
        "Start",
        " ",
        "Time:",
        " ",
        "11/02/2025",
        " ",
        "06:30",
      ],
    },
    {
      body: [
        "Session",
        " ",
        "End",
        " ",
        "Time:",
        " ",
        "11/02/2025",
        " ",
        "09:11",
      ],
    },
    {
      body: ["Session", " ", "Hrs:", " ", "2.68"],
    },
    {
      body: ["Role:", " ", "Picker"],
    },
    {
      body: ["Individual", " ", "Pick", " ", "Records"],
    },
    {
      body: ["07:13", " ", "38.00"],
    },
    {
      body: ["07:49", " ", "40.00"],
    },
    {
      body: ["08:30", " ", "42.00"],
    },
    {
      body: ["08:53", " ", "16.00"],
    },
    {
      body: ["No.", " ", "of", " ", "Pick", " ", "Records:", " ", "4"],
    },
    {
      body: ["Subtotal", " ", "Pick", " ", "Records:", " ", "136.0"],
    },
    {
      body: ["Actual", " ", "Pick", " ", "Rate:", " ", "50.75", ""],
    },
    {
      body: ["Break", ""],
    },
    {
      body: ["Break", " ", "Type:", " ", "Paid", " ", "Break"],
    },
    {
      body: ["Start", " ", "Time:", " ", "11/02/2025", " ", "09:11"],
    },
    {
      body: ["End", " ", "Time:", " ", "11/02/2025", " ", "09:26"],
    },
    {
      body: ["Break", " ", "Hrs:", " ", "0.25", ""],
    },
    {
      body: ["Session", ""],
    },
    {
      body: ["Location:", " ", "[Bengeo]609"],
    },
    {
      body: ["Variety:", " ", "Victoria"],
    },
    {
      body: ["Pay", " ", "Type:", " ", "Per", " ", "Unit"],
    },
    {
      body: ["Unit:", " ", "125g", " ", "Driscoll's"],
    },
    {
      body: ["Piece", " ", "Rate:", " ", "0.71"],
    },
    {
      body: ["Target", " ", "Pick", " ", "Rate:", " ", "48.00"],
    },
    {
      body: [
        "Session",
        " ",
        "Start",
        " ",
        "Time:",
        " ",
        "11/02/2025",
        " ",
        "09:26",
      ],
    },
    {
      body: [
        "Session",
        " ",
        "End",
        " ",
        "Time:",
        " ",
        "11/02/2025",
        " ",
        "12:00",
      ],
    },
    {
      body: ["Session", " ", "Hrs:", " ", "2.57"],
    },
    {
      body: ["Role:", " ", "Picker"],
    },
    {
      body: ["Individual", " ", "Pick", " ", "Records"],
    },
    {
      body: ["09:50", " ", "39.00"],
    },
    {
      body: ["10:20", " ", "38.00"],
    },
    {
      body: ["10:52", " ", "40.00"],
    },
    {
      body: ["11:29", " ", "39.00"],
    },
    {
      body: ["11:54", " ", "20.00"],
    },
    {
      body: ["No.", " ", "of", " ", "Pick", " ", "Records:", " ", "5"],
    },
    {
      body: ["Subtotal", " ", "Pick", " ", "Records:", " ", "176.0"],
    },
    {
      body: ["Actual", " ", "Pick", " ", "Rate:", " ", "68.48", ""],
    },
    {
      body: ["Totals", ""],
    },
    {
      body: [
        "Total",
        " ",
        "Pick",
        " ",
        "Records",
        " ",
        "by",
        " ",
        "-",
        " ",
        "125g",
        " ",
        "Driscoll's:",
        " ",
        "312.00",
      ],
    },
    {
      body: ["Total", " ", "Session", " ", "Hrs:", " ", "5.25"],
    },
    {
      body: ["Total", " ", "Picker", " ", "Hrs:", " ", "5.25"],
    },
    {
      body: ["Total", " ", "Paid", " ", "Break", " ", "Hrs:", " ", "0.25"],
    },
    {
      body: ["Total", " ", "Paid", " ", "Hrs:", " ", "5.50"],
    },
    {
      body: ["Avg", " ", "Pick", " ", "Rate:", " ", "59.61"],
    },
    {
      body: ["Supervisor:", " ", "Michael,", " ", "Jackson"],
    },
  ],
};

describe("normalize lines", () => {
  test("case1", () => {
    expect(composeSummaryReceipt(testData.lines)).toMatchSnapshot();
  });
});
