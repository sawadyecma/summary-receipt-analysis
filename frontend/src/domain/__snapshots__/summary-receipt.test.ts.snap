// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`normalize lines > case1 1`] = `
{
  "body": [
    {
      "body": [
        "Name: Bubbles, Jackson",
        "Employee ID: 126167",
        "Pick Date: 11/02/2025",
        "Start Time: 11/02/2025 06:30",
        "End Time: 11/02/2025 12:02",
        "Attendance Hrs: 5.53",
      ],
      "sectionType": "header",
    },
    {
      "body": [
        "Session",
        "Location: [Bengeo]608",
        "Variety: Victoria",
        "Pay Type: Per Unit",
        "Unit: 125g Driscoll's",
        "Piece Rate: 0.71",
        "Target Pick Rate: 48.00",
        "Session Start Time: 11/02/2025 06:30",
        "Session End Time: 11/02/2025 09:11",
        "Session Hrs: 2.68",
        "Role: Picker",
        "Individual Pick Records",
        "07:13 38.00",
        "07:49 40.00",
        "08:30 42.00",
        "08:53 16.00",
        "No. of Pick Records: 4",
        "Subtotal Pick Records: 136.0",
        "Actual Pick Rate: 50.75",
      ],
      "sectionType": "session",
    },
    {
      "body": [
        "Break",
        "Break Type: Paid Break",
        "Start Time: 11/02/2025 09:11",
        "End Time: 11/02/2025 09:26",
        "Break Hrs: 0.25",
      ],
      "sectionType": "break",
    },
    {
      "body": [
        "Session",
        "Location: [Bengeo]609",
        "Variety: Victoria",
        "Pay Type: Per Unit",
        "Unit: 125g Driscoll's",
        "Piece Rate: 0.71",
        "Target Pick Rate: 48.00",
        "Session Start Time: 11/02/2025 09:26",
        "Session End Time: 11/02/2025 12:00",
        "Session Hrs: 2.57",
        "Role: Picker",
        "Individual Pick Records",
        "09:50 39.00",
        "10:20 38.00",
        "10:52 40.00",
        "11:29 39.00",
        "11:54 20.00",
        "No. of Pick Records: 5",
        "Subtotal Pick Records: 176.0",
        "Actual Pick Rate: 68.48",
      ],
      "sectionType": "session",
    },
    {
      "body": [
        "Totals",
        "Total Pick Records by - 125g Driscoll's: 312.00",
        "Total Session Hrs: 5.25",
        "Total Picker Hrs: 5.25",
        "Total Paid Break Hrs: 0.25",
        "Total Paid Hrs: 5.50",
        "Avg Pick Rate: 59.61",
        "Supervisor: Michael, Jackson",
      ],
      "sectionType": "totals",
    },
  ],
}
`;
